lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,493 @@
+//===-- Utils.cpp - linalgx utils -------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Parser/Parser.h"
+
+#include "gc/Dialect/Linalgx/Utils.h"
+
+namespace mlir {
+namespace linalgx {
+
+/// Packed Matmul Structs
+struct BatchDimMap {
+public:
+  BatchDimMap() = default;
+  BatchDimMap(ArrayRef<int64_t> batchARef, ArrayRef<int64_t> batchBRef,
+              ArrayRef<int64_t> batchCRef)
+      : batchA(batchARef), batchB(batchBRef), batchC(batchCRef) {}
+  /// Get original arrays
+  ArrayRef<int64_t> getBatchA() const { return ArrayRef<int64_t>(batchA); }
+  ArrayRef<int64_t> getBatchB() const { return ArrayRef<int64_t>(batchB); }
+  ArrayRef<int64_t> getBatchC() const { return ArrayRef<int64_t>(batchC); }
+  /// Get size of batch dims
+  int64_t getBatchNum() const { return batchA.size(); }
+  /// Get attr for batch info
+  bool isBatchEmpty() const {
+    return batchA.empty() && batchB.empty() && batchC.empty();
+  }
+  bool isBatchMatmul() const {
+    return !isBatchEmpty() && batchEqualAB() && batchEqualAC();
+  }
+  bool isBatchReduce() const {
+    return !isBatchEmpty() && batchEqualAB() && batchC.empty();
+  }
+
+private:
+  bool batchEqualAB() const {
+    return std::equal(batchA.begin(), batchA.end(), batchB.begin());
+  }
+  bool batchEqualAC() const {
+    return std::equal(batchA.begin(), batchA.end(), batchC.begin());
```

Comment:
Please check batchA and batchC has the same size otherwise it may throw segfault

Reply:
Switch to `llvm::equal` which uses C++14 `std::equal` and check for 2 ranges.